File: apps/minecraft/spigot/src/test/java/crabcraft/net/crabUtilities/voicechat/LofiVoicechatRegressionTest.java
package crabcraft.net.crabUtilities.voicechat;

import de.maxhenkel.voicechat.api.Group;
import de.maxhenkel.voicechat.api.ServerPlayer;
import de.maxhenkel.voicechat.api.VoicechatConnection;
import de.maxhenkel.voicechat.api.VoicechatServerApi;
import de.maxhenkel.voicechat.api.audiochannel.StaticAudioChannel;
import de.maxhenkel.voicechat.api.events.SoundPacketEvent;
import net.crabcraft.customdiscs.Keys;

import java.lang.reflect.Proxy;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.logging.Logger;

final class LofiVoicechatRegressionTest {
    public static void main(String[] args) throws Exception {
        verifyGroupSelection();
        verifySpeechRoutingSelection();
        verifyVolumeScaling();
        verifyLofiChannelTargets();
        verifyDefaultsAndLegacyNamespace();
    }

    private static void verifyGroupSelection() {
        List<String> enabled = CrabVoicechatPlugin.persistentGroupNames(List.of("Global #1"), true);
        check(enabled.equals(List.of("Global #1", "24/7 Lofi")),
                "enabled lofi group was not added exactly once");
        List<String> disabled = CrabVoicechatPlugin.persistentGroupNames(
                List.of("Global #1", "24/7 Lofi"), false);
        check(disabled.equals(List.of("Global #1")), "disabled lofi group was still created");

        UUID first = CrabVoicechatPlugin.deterministicGroupId("24/7 Lofi");
        UUID second = CrabVoicechatPlugin.deterministicGroupId("24/7 Lofi");
        check(first.equals(second), "lofi group ID was not deterministic across backends");
    }

    private static void verifySpeechRoutingSelection() {
        UUID lofiId = CrabVoicechatPlugin.deterministicGroupId("24/7 Lofi");
        Group lofiGroup = proxy(Group.class, (proxy, method, args) -> switch (method.getName()) {
            case "getId" -> lofiId;
            default -> throw new AssertionError("unexpected group call: " + method.getName());
        });
        VoicechatConnection sender = proxy(VoicechatConnection.class,
                (proxy, method, args) -> switch (method.getName()) {
                    case "getGroup" -> lofiGroup;
                    default -> throw new AssertionError("unexpected connection call: " + method.getName());
                });

        check(GroupSpeechAttenuator.shouldAttenuate(SoundPacketEvent.SOURCE_GROUP, lofiId, sender),
                "lofi group speech was not selected for attenuation");
        check(!GroupSpeechAttenuator.shouldAttenuate(SoundPacketEvent.SOURCE_PROXIMITY, lofiId, sender),
                "proximity speech was incorrectly selected for attenuation");
        check(!GroupSpeechAttenuator.shouldAttenuate(
                        SoundPacketEvent.SOURCE_GROUP, UUID.randomUUID(), sender),
                "another group was incorrectly selected for attenuation");
    }

    private static void verifyVolumeScaling() {
        short[] speech = {10_000, -10_000, Short.MAX_VALUE, Short.MIN_VALUE};
        OpusVolumeScaler.applyGain(speech, 0.25D);
        check(speech[0] == 2_500 && speech[1] == -2_500,
                "75% speech reduction did not produce a 25% signal");
        check(speech[2] == 8_192 && speech[3] == -8_192,
                "speech volume scaling changed endpoint rounding");
    }

    private static void verifyLofiChannelTargets() {
        UUID lofiId = CrabVoicechatPlugin.deterministicGroupId("24/7 Lofi");
        AtomicInteger added = new AtomicInteger();
        AtomicInteger removed = new AtomicInteger();
        AtomicInteger cleared = new AtomicInteger();

        StaticAudioChannel channel = proxy(StaticAudioChannel.class, (proxy, method, args) -> {
            switch (method.getName()) {
                case "addTarget" -> added.incrementAndGet();
                case "removeTarget" -> removed.incrementAndGet();
                case "clearTargets" -> cleared.incrementAndGet();
                case "setBypassGroupIsolation", "setCategory", "setFilter", "flush" -> { }
                case "bypassesGroupIsolation" -> { return true; }
                default -> throw new AssertionError("unexpected channel call: " + method.getName());
            }
            return null;
        });
        VoicechatServerApi api = proxy(VoicechatServerApi.class, (proxy, method, args) -> {
            if (method.getName().equals("createStaticAudioChannel") && args.length == 1) return channel;
            throw new AssertionError("unexpected API call: " + method.getName());
        });

        UUID playerId = UUID.randomUUID();
        Group lofiGroup = group(lofiId);
        VoicechatConnection connection = connection(playerId, lofiGroup);
        LofiStreamPlayer player = new LofiStreamPlayer(
                api, lofiId, "https://example.invalid/live", 0.5F,
                Logger.getLogger("LofiVoicechatRegressionTest"));

        check(player.openChannel(), "lofi static channel was not opened");
        player.reconcileTarget(connection);
        check(added.get() == 1, "local lofi member was not registered as an audio target");

        player.updateTarget(connection, UUID.randomUUID());
        check(removed.get() == 1, "player leaving lofi was not removed as an audio target");

        player.updateTarget(connection, lofiId);
        player.removeTarget(playerId);
        check(added.get() == 2 && removed.get() == 2,
                "disconnect did not remove the tracked lofi audio target");

        player.close();
        check(cleared.get() == 1, "lofi targets were not cleared during shutdown");
    }

    private static Group group(UUID id) {
        return proxy(Group.class, (proxy, method, args) -> {
            if (method.getName().equals("getId")) return id;
            throw new AssertionError("unexpected group call: " + method.getName());
        });
    }

    private static VoicechatConnection connection(UUID playerId, Group group) {
        ServerPlayer serverPlayer = proxy(ServerPlayer.class, (proxy, method, args) -> {
            if (method.getName().equals("getUuid")) return playerId;
            throw new AssertionError("unexpected player call: " + method.getName());
        });
        return proxy(VoicechatConnection.class, (proxy, method, args) -> switch (method.getName()) {
            case "getPlayer" -> serverPlayer;
            case "getGroup" -> group;
            default -> throw new AssertionError("unexpected connection call: " + method.getName());
        });
    }

    private static void verifyDefaultsAndLegacyNamespace() throws Exception {
        String config;
        try (var input = LofiVoicechatRegressionTest.class.getClassLoader().getResourceAsStream("config.yml")) {
            check(input != null, "bundled config.yml is missing");
            config = new String(input.readAllBytes(), StandardCharsets.UTF_8);
        }
        check(config.contains("https://www.youtube.com/watch?v=8TIVlFtcRDU"),
                "configured YouTube Live default changed");
        check(config.contains("music-volume: 0.5") && config.contains("player-volume: 0.25"),
                "lofi volume defaults changed");
        check(Keys.REMOTE_DISC.key().getNamespace().equals("customdiscs"),
                "legacy custom-disc item namespace changed during the merge");
    }

    @SuppressWarnings("unchecked")
    private static <T> T proxy(Class<T> type, java.lang.reflect.InvocationHandler handler) {
        return (T) Proxy.newProxyInstance(type.getClassLoader(), new Class<?>[]{type}, handler);
    }

    private static void check(boolean condition, String message) {
        if (!condition) throw new AssertionError(message);
    }
}

File: apps/minecraft/spigot/src/test/kotlin/crabcraft/net/crabUtilities/voicechat/LofiVoicechatRegressionTest.kt
package crabcraft.net.crabUtilities.voicechat

import de.maxhenkel.voicechat.api.Group
import de.maxhenkel.voicechat.api.ServerPlayer
import de.maxhenkel.voicechat.api.VoicechatConnection
import de.maxhenkel.voicechat.api.VoicechatServerApi
import de.maxhenkel.voicechat.api.audiochannel.StaticAudioChannel
import de.maxhenkel.voicechat.api.events.SoundPacketEvent
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Proxy
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger
import java.util.logging.Logger

object LofiVoicechatRegressionTest {
    @JvmStatic
    fun main(args: Array<String>) {
        verifyGroupSelection()
        verifySpeechRoutingSelection()
        verifyVolumeScaling()
        verifyLofiChannelTargets()
        check(
            LofiStreamPlayer.RETRY_SECONDS >= 300L,
            "lofi resolution failures would retry often enough to flood the console",
        )
    }

    private fun verifyGroupSelection() {
        val enabled = CrabVoicechatPlugin.persistentGroupNames(listOf("Global #1", "Global #2", "Global #3"), true)
        check(
            enabled == listOf("Global #1", "Global #2", "Global #3", CrabVoicechatPlugin.LOFI_GROUP_NAME),
            "enabled lofi group was not added after the three global groups",
        )
        check(enabled.sorted() == enabled, "Simple Voice Chat's alphabetical ordering would not place lofi fourth")
        val disabled =
            CrabVoicechatPlugin.persistentGroupNames(
                listOf("Global #1", "24/7 Lofi", CrabVoicechatPlugin.LOFI_GROUP_NAME),
                false,
            )
        check(disabled == listOf("Global #1"), "disabled lofi group was still created")
        check(
            CrabVoicechatPlugin.deterministicGroupId("24/7 Lofi") ==
                CrabVoicechatPlugin.deterministicGroupId("24/7 Lofi"),
            "lofi group ID was not deterministic across backends",
        )
        check(
            CrabVoicechatPlugin.LOFI_GROUP_NAME.length == 16,
            "lofi group name exceeds Simple Voice Chat's plugin API limit",
        )
    }

    private fun verifySpeechRoutingSelection() {
        val lofiId = CrabVoicechatPlugin.deterministicGroupId("24/7 Lofi")
        val lofiGroup = group(lofiId)
        val sender =
            proxy(VoicechatConnection::class.java) { _, method, _ ->
                if (method.name == "getGroup") lofiGroup
                else throw AssertionError("unexpected connection call: ${method.name}")
            }
        check(
            GroupSpeechAttenuator.shouldAttenuate(SoundPacketEvent.SOURCE_GROUP, lofiId, sender),
            "lofi group speech was not selected for attenuation",
        )
        check(
            !GroupSpeechAttenuator.shouldAttenuate(SoundPacketEvent.SOURCE_PROXIMITY, lofiId, sender),
            "proximity speech was incorrectly selected for attenuation",
        )
        check(
            !GroupSpeechAttenuator.shouldAttenuate(SoundPacketEvent.SOURCE_GROUP, UUID.randomUUID(), sender),
            "another group was incorrectly selected for attenuation",
        )
    }

    private fun verifyVolumeScaling() {
        val speech = shortArrayOf(10_000, -10_000, Short.MAX_VALUE, Short.MIN_VALUE)
        OpusVolumeScaler.applyGain(speech, 0.25)
        check(
            speech[0] == 2_500.toShort() && speech[1] == (-2_500).toShort(),
            "75% speech reduction did not produce a 25% signal",
        )
        check(
            speech[2] == 8_192.toShort() && speech[3] == (-8_192).toShort(),
            "speech volume scaling changed endpoint rounding",
        )
    }

    private fun verifyLofiChannelTargets() {
        val lofiId = CrabVoicechatPlugin.deterministicGroupId("24/7 Lofi")
        val added = AtomicInteger()
        val removed = AtomicInteger()
        val cleared = AtomicInteger()
        val channel =
            proxy(StaticAudioChannel::class.java) { _, method, _ ->
                when (method.name) {
                    "addTarget" -> {
                        added.incrementAndGet()
                        null
                    }
                    "removeTarget" -> {
                        removed.incrementAndGet()
                        null
                    }
                    "clearTargets" -> {
                        cleared.incrementAndGet()
                        null
                    }
                    "setBypassGroupIsolation",
                    "setCategory",
                    "setFilter",
                    "flush" -> null
                    "bypassesGroupIsolation" -> true
                    else -> throw AssertionError("unexpected channel call: ${method.name}")
                }
            }
        val api =
            proxy(VoicechatServerApi::class.java) { _, method, args ->
                if (method.name == "createStaticAudioChannel" && args!!.size == 1) channel
                else throw AssertionError("unexpected API call: ${method.name}")
            }
        val playerId = UUID.randomUUID()
        val connection = connection(playerId, group(lofiId))
        val player =
            LofiStreamPlayer(
                api,
                lofiId,
                "https://example.invalid/live",
                0.5F,
                Logger.getLogger("LofiVoicechatRegressionTest"),
            )
        check(player.openChannel(), "lofi static channel was not opened")
        player.reconcileTarget(connection)
        check(added.get() == 1, "local lofi member was not registered as an audio target")
        player.updateTarget(connection, UUID.randomUUID())
        check(removed.get() == 1, "player leaving lofi was not removed as an audio target")
        player.updateTarget(connection, lofiId)
        player.removeTarget(playerId)
        check(added.get() == 2 && removed.get() == 2, "disconnect did not remove the tracked lofi audio target")
        player.close()
        check(cleared.get() == 1, "lofi targets were not cleared during shutdown")
    }

    private fun group(id: UUID) =
        proxy(Group::class.java) { _, method, _ ->
            if (method.name == "getId") id else throw AssertionError("unexpected group call: ${method.name}")
        }

    private fun connection(playerId: UUID, group: Group): VoicechatConnection {
        val player =
            proxy(ServerPlayer::class.java) { _, method, _ ->
                if (method.name == "getUuid") playerId
                else throw AssertionError("unexpected player call: ${method.name}")
            }
        return proxy(VoicechatConnection::class.java) { _, method, _ ->
            when (method.name) {
                "getPlayer" -> player
                "getGroup" -> group
                else -> throw AssertionError("unexpected connection call: ${method.name}")
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> proxy(type: Class<T>, handler: InvocationHandler): T =
        Proxy.newProxyInstance(type.classLoader, arrayOf(type), handler) as T

    private fun check(condition: Boolean, message: String) {
        if (!condition) throw AssertionError(message)
    }
}
